### ISSUE TITLE
MB_ENCRYPTION_SECRET_KEY after upgrade from 0.31.2 to 0.32.3

### ISSUE THREAD

**Issue by SirTomme** [metabase/metabase#9737]
Your browser and the version: Google Chrome 71.0.3578.80 but also happens in Firefox 52 ESR
Your operating system: Windows 7
Your databases: MySQL, Oracle
Metabase version: 0.32.3
Metabase hosting environment: RHEL 6.6
Metabase internal database: MySQL 5.6

Maybe this is related to the closed issue #8151

After upgrading from 0.31.2 to 0.32.3, when saving a modified question that was created before the upgrade, I get the error shown below. This error is reproducable with some other questions that existed before the upgrade. I just changed the formatting of a line chart because it was changed (reset to defaults) during the upgrade process.
The questions only use MySQL and the error persists after re-entering the password for the external database in the admin panel although the changes are being saved now.

EDIT: I should mention that I did not change the MB_ENCRYPTION_SECRET_KEY. It was kept untouched in /etc/default/metabase (which I referenced in /etc/init.d/metabase).

```
04-09 15:02:00 WARN util.encryption :: Konnte verschl?sselte Strings nicht entschl?sseln. Hast du vergessen MB_ENCRYPTION_SECRET_KEY zu setzen oder den Wert ge?ndert? 0 > -32 ("java.util.Arrays.copyOfRange(Arrays.java:3519)"
 "buddy.core.bytes$slice.invokeStatic(bytes.clj:47)"
 "buddy.core.bytes$slice.invokePrim(bytes.clj)"
 "buddy.core.crypto$fn__21432.invokeStatic(crypto.clj:485)"
 "buddy.core.crypto$fn__21432.invoke(crypto.clj:478)"
 "clojure.lang.MultiFn.invoke(MultiFn.java:229)"
 "buddy.core.crypto$decrypt.invokeStatic(crypto.clj:572)"
 "buddy.core.crypto$decrypt.invoke(crypto.clj:559)"
 "--> util.encryption$decrypt.invokeStatic(encryption.clj:65)"
 "util.encryption$decrypt.invoke(encryption.clj:57)"
 "util.encryption$maybe_decrypt.invokeStatic(encryption.clj:103)"
 "util.encryption$maybe_decrypt.invoke(encryption.clj:96)"
 "util.encryption$maybe_decrypt.invokeStatic(encryption.clj:99)"
 "util.encryption$maybe_decrypt.invoke(encryption.clj:96)"
 "query_processor.middleware.results_metadata$valid_checksum_QMARK_.invokeStatic(results_metadata.clj:79)"
 "query_processor.middleware.results_metadata$valid_checksum_QMARK_.invoke(results_metadata.clj:74)"
 "api.card$fn__43819$result_metadata_async__43824$fn__43825.invoke(card.clj:194)"
 "api.card$fn__43819$result_metadata_async__43824.invoke(card.clj:189)"
 "api.card$result_metadata_for_updating_async.invokeStatic(card.clj:308)"
 "api.card$result_metadata_for_updating_async.invoke(card.clj:298)"
 "api.card$fn__44053.invokeStatic(card.clj:454)"
 "api.card$fn__44053.invoke(card.clj:430)"
 "middleware.auth$enforce_authentication$fn__59520.invoke(auth.clj:14)"
 "routes$fn__61158$fn__61159.doInvoke(routes.clj:113)"
 "middleware.exceptions$catch_uncaught_exceptions$fn__59611.invoke(exceptions.clj:104)"
 "middleware.exceptions$catch_api_exceptions$fn__59608.invoke(exceptions.clj:92)"
 "middleware.log$log_api_call$fn__61038$fn__61039.invoke(log.clj:168)"
 "middleware.log$log_api_call$fn__61038.invoke(log.clj:162)"
 "middleware.security$add_security_headers$fn__59574.invoke(security.clj:102)"
 "middleware.json$wrap_json_body$fn__60743.invoke(json.clj:61)"
 "middleware.json$wrap_streamed_json_response$fn__60761.invoke(json.clj:97)"
 "middleware.session$bind_current_user$fn__58230$fn__58231.invoke(session.clj:185)"
 "middleware.session$do_with_current_user.invokeStatic(session.clj:168)"
 "middleware.session$do_with_current_user.invoke(session.clj:162)"
 "middleware.session$bind_current_user$fn__58230.invoke(session.clj:184)"
 "middleware.session$wrap_current_user_id$fn__58219.invoke(session.clj:153)"
 "middleware.session$wrap_session_id$fn__58204.invoke(session.clj:115)"
 "middleware.auth$wrap_api_key$fn__59528.invoke(auth.clj:27)"
 "middleware.misc$maybe_set_site_url$fn__61101.invoke(misc.clj:54)"
 "middleware.misc$bind_user_locale$fn__61104.invoke(misc.clj:70)"
 "middleware.misc$add_content_type$fn__61091.invoke(misc.clj:29)"
 "middleware.misc$wrap_gzip$fn__61119.invoke(misc.clj:97)")

04-09 15:02:00 WARN util.encryption :: Konnte verschl?sselte Strings nicht entschl?sseln. Hast du vergessen MB_ENCRYPTION_SECRET_KEY zu setzen oder den Wert ge?ndert? 0 > -32 ("java.util.Arrays.copyOfRange(Arrays.java:3519)"
 "buddy.core.bytes$slice.invokeStatic(bytes.clj:47)"
 "buddy.core.bytes$slice.invokePrim(bytes.clj)"
 "buddy.core.crypto$fn__21432.invokeStatic(crypto.clj:485)"
 "buddy.core.crypto$fn__21432.invoke(crypto.clj:478)"
 "clojure.lang.MultiFn.invoke(MultiFn.java:229)"
 "buddy.core.crypto$decrypt.invokeStatic(crypto.clj:572)"
 "buddy.core.crypto$decrypt.invoke(crypto.clj:559)"
 "--> util.encryption$decrypt.invokeStatic(encryption.clj:65)"
 "util.encryption$decrypt.invoke(encryption.clj:57)"
 "util.encryption$maybe_decrypt.invokeStatic(encryption.clj:103)"
 "util.encryption$maybe_decrypt.invoke(encryption.clj:96)"
 "util.encryption$maybe_decrypt.invokeStatic(encryption.clj:99)"
 "util.encryption$maybe_decrypt.invoke(encryption.clj:96)"
 "query_processor.middleware.results_metadata$valid_checksum_QMARK_.invokeStatic(results_metadata.clj:80)"
 "query_processor.middleware.results_metadata$valid_checksum_QMARK_.invoke(results_metadata.clj:74)"
 "api.card$fn__43819$result_metadata_async__43824$fn__43825.invoke(card.clj:194)"
 "api.card$fn__43819$result_metadata_async__43824.invoke(card.clj:189)"
 "api.card$result_metadata_for_updating_async.invokeStatic(card.clj:308)"
 "api.card$result_metadata_for_updating_async.invoke(card.clj:298)"
 "api.card$fn__44053.invokeStatic(card.clj:454)"
 "api.card$fn__44053.invoke(card.clj:430)"
 "middleware.auth$enforce_authentication$fn__59520.invoke(auth.clj:14)"
 "routes$fn__61158$fn__61159.doInvoke(routes.clj:113)"
 "middleware.exceptions$catch_uncaught_exceptions$fn__59611.invoke(exceptions.clj:104)"
 "middleware.exceptions$catch_api_exceptions$fn__59608.invoke(exceptions.clj:92)"
 "middleware.log$log_api_call$fn__61038$fn__61039.invoke(log.clj:168)"
 "middleware.log$log_api_call$fn__61038.invoke(log.clj:162)"
 "middleware.security$add_security_headers$fn__59574.invoke(security.clj:102)"
 "middleware.json$wrap_json_body$fn__60743.invoke(json.clj:61)"
 "middleware.json$wrap_streamed_json_response$fn__60761.invoke(json.clj:97)"
 "middleware.session$bind_current_user$fn__58230$fn__58231.invoke(session.clj:185)"
 "middleware.session$do_with_current_user.invokeStatic(session.clj:168)"
 "middleware.session$do_with_current_user.invoke(session.clj:162)"
 "middleware.session$bind_current_user$fn__58230.invoke(session.clj:184)"
 "middleware.session$wrap_current_user_id$fn__58219.invoke(session.clj:153)"
 "middleware.session$wrap_session_id$fn__58204.invoke(session.clj:115)"
 "middleware.auth$wrap_api_key$fn__59528.invoke(auth.clj:27)"
 "middleware.misc$maybe_set_site_url$fn__61101.invoke(misc.clj:54)"
 "middleware.misc$bind_user_locale$fn__61104.invoke(misc.clj:70)"
 "middleware.misc$add_content_type$fn__61091.invoke(misc.clj:29)"
 "middleware.misc$wrap_gzip$fn__61119.invoke(misc.clj:97)")

04-09 15:02:00 INFO api.card :: Die an die API ?bergebenen Metatdaten zum Ergebnis der Karte sind G?LTIG. Vielen Dank!
```

**Comment by reidab** [comment]
I'm seeing the same issue after upgrading one minor patch release from 0.32.2 to 0.32.3 with no changes other than changing the docker tag. On server startup I see the following stacktrace and all attempts to auth using Google OAuth SSO fail.

```
04-09 06:58:40 [WARN util.encryption]:: Cannot decrypt encrypted string. Have you changed or forgot to set MB_ENCRYPTION_SECRET_KEY? Message seems corrupt or manipulated. ("buddy.core.crypto$fn__21432.invokeStatic(crypto.clj:489)"
 "buddy.core.crypto$fn__21432.invoke(crypto.clj:478)"
 "clojure.lang.MultiFn.invoke(MultiFn.java:229)"
 "buddy.core.crypto$decrypt.invokeStatic(crypto.clj:572)"
 "buddy.core.crypto$decrypt.invoke(crypto.clj:559)"
 "--> util.encryption$decrypt.invokeStatic(encryption.clj:65)"
 "util.encryption$decrypt.invoke(encryption.clj:57)"
 "util.encryption$maybe_decrypt.invokeStatic(encryption.clj:103)"
 "util.encryption$maybe_decrypt.invoke(encryption.clj:96)"
 "util.encryption$maybe_decrypt.invokeStatic(encryption.clj:99)"
 "util.encryption$maybe_decrypt.invoke(encryption.clj:96)"
 "models.setting.cache$restore_cache_if_needed_BANG_.invokeStatic(cache.clj:133)"
 "models.setting.cache$restore_cache_if_needed_BANG_.invoke(cache.clj:117)"
 "models.setting$db_value.invokeStatic(setting.clj:130)"
 "models.setting$db_value.invoke(setting.clj:124)"
 "models.setting$get_string.invokeStatic(setting.clj:144)"
 "models.setting$get_string.invoke(setting.clj:134)"
 "models.setting$get.invokeStatic(setting.clj:204)"
 "models.setting$get.invoke(setting.clj:198)"
 "core$init_BANG_.invokeStatic(core.clj:105)"
 "core$init_BANG_.invoke(core.clj:51)"
 "core$start_normally.invokeStatic(core.clj:118)"
 "core$start_normally.invoke(core.clj:112)"
 "core$_main.invokeStatic(core.clj:138)"
 "core$_main.doInvoke(core.clj:133)"
 "core.main(Unknown Source)")
```

**Comment by jazarja** [comment]
I got same message on my embedded chart even tough I haven't update metabase server version. I use metabase version v0.31.2

**Comment by Bacto** [comment]
Hi,

I have the same error with v0.32.10 as v0.33.0-preview2 (I haven't test older ones).

To reproduce the bug, simply start Metabase with and encryption key.
I've used the one provided in the [start guide](https://metabase.com/docs/latest/operations-guide/start.html#encrypting-your-database-connection-details-at-rest):
`docker run -p 3000:3000 -e MB_ENCRYPTION_SECRET_KEY="IYqrSi5QDthvFWe4/WdAxhnra5DZC3RKx3ZSrOJDKsM=" metabase/metabase:v0.32.10`

Then create an account and click on "Sample Dataset" / "Orders" / "Save" / "Save".
You will get this error: `WARN util.encryption :: Cannot decrypt encrypted string. Have you changed or forgot to set MB_ENCRYPTION_SECRET_KEY? 0 > -32 ("java.base/java.util.Arrays.copyOfRange(Unknown Source)"`

What is strange is that all seems working well.

**Comment by camsaul** [comment]
Fixed by #10539
